master/internal/api_generic_tasks.go

```diff
@@ -0,0 +1,761 @@
+	}
+
+	// Validate the resource configuration.
+	resources := model.ParseJustResources(configBytes)
+
+	if resources.Slots < 0 {
+		return nil, nil, nil, fmt.Errorf("resource slots must be >= 0")
+	}
+	poolName, launchWarnings, err := a.m.ResolveResources(resources.ResourcePool, resources.Slots, int(proj.WorkspaceId))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	// Get the base TaskSpec.
+	taskSpec, err := a.m.fillTaskSpec(poolName, agentUserGroup, userModel)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// Get the full configuration.
+	taskConfig := model.DefaultConfigGenericTaskConfig(&taskSpec.TaskContainerDefaults)
+	if err := yaml.UnmarshalStrict(configBytes, &taskConfig, yaml.DisallowUnknownFields); err != nil {
+		return nil, nil, nil, fmt.Errorf("yaml unmarshaling generic task config: %w", err)
+	}
+	workDirInDefaults := taskConfig.WorkDir
+
+	// Copy discovered (default) resource pool name and slot count.
+
+	fillTaskConfig(resources.Slots, taskSpec, &taskConfig.Environment)
+	taskConfig.Resources.RawResourcePool = &poolName
+	taskConfig.Resources.RawSlots = &resources.Slots
+
+	var contextDirectoryBytes []byte
+	taskConfig.WorkDir, contextDirectoryBytes, err = fillContextDir(
+		taskConfig.WorkDir,
+		workDirInDefaults,
+		contextDirectory,
+	)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	var token string
+	token, err = getTaskSessionToken(ctx, userModel)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	taskSpec.UserSessionToken = token
+
+	genericTaskSpec.Base = taskSpec
+	genericTaskSpec.GenericTaskConfig = taskConfig
+
+	genericTaskSpec.Base.ExtraEnvVars = map[string]string{
+		"DET_TASK_TYPE": string(model.TaskTypeGeneric),
+	}
+
+	return genericTaskSpec, launchWarnings, contextDirectoryBytes, nil
+}
+
+func (a *apiServer) canCreateGenericTask(ctx context.Context, projectID int) error {
+	userModel, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return err
+	}
+
+	errProjectNotFound := api.NotFoundErrs("project", fmt.Sprint(projectID), true)
+	p := &projectv1.Project{}
+	// Get project details
+	if err := db.Bun().NewSelect().TableExpr("pe, projects AS p").With("pe", db.Bun().NewSelect().
+		ModelTableExpr("experiments").ColumnExpr(`COUNT(*) AS num_experiments,
+		SUM(CASE WHEN state = 'ACTIVE' THEN 1 ELSE 0 END) AS num_active_experiments,
+		MAX(start_time) AS last_experiment_started_at`).Where("project_id = ?", projectID)).
+		ColumnExpr(`p.id,p.name,p.workspace_id,p.description,p.immutable,
+		p.notes,w.name AS workspace_name,p.error_message,
+		(p.archived OR w.archived) AS archived,MAX(pe.num_experiments) AS num_experiments,
+		MAX(pe.num_active_experiments) AS num_active_experiments,u.username,p.user_id`).
+		Join("LEFT JOIN users AS u ON u.id = p.user_id").
+		Join("LEFT JOIN workspaces AS w ON w.id = p.workspace_id").
+		Where("p.id = ?", projectID).
+		GroupExpr("p.id, u.username, w.archived, w.name").
+		Scan(ctx, p); errors.Is(err, db.ErrNotFound) {
+		return errProjectNotFound
```

Comment:
this formatting style is a bit tough to read. can I suggest
1. indenting it with each function call (`ColumnExpr`, `Where`) starting new lines, like here: https://github.com/determined-ai/determined/blob/main/master/internal/api_experiment.go#L2629
2. breaking out the subquery under `.With` into another variable